docs/design/018-new-http-api.md

```diff
@@ -0,0 +1,486 @@
+
+It can be of three types:
+- `tcp`: a stream-oriented network socket using the Transmission Control Protocol.
+- `udp`: a message-oriented network socket using the User Datagram Protocol.
+- `ipc`: a mechanism for communicating between processes on the same machine, typically using files.
+
+The Socket state can either be _active_—meaning connected for a TCP socket, bound for a UDP socket, or open for an IPC socket—, or _inactive_—meaning disconnected, unbound, or closed, respectively.
+
+##### Example
+
+- TCP:
+```javascript
+import { TCP } from 'k6/x/net';
+
+export default async function () {
+  const socket = await TCP.open('192.168.1.1:80', {
+                            // default      | possible values
+    ipVersion: 0,           // 0            | 4 (IPv4), 6 (IPv6), 0 (both)
+    keepAlive: true,        // false        |
+    lookup: null,           // dns.lookup() |
+    proxy: 'myproxy:3030',  // ''           |
+  });
+  console.log(socket.active); // true
+
+  // Writing directly to the socket.
+  // Requires TextEncoder implementation, otherwise typed arrays can be used as well.
+  await socket.write(new TextEncoder().encode('GET / HTTP/1.1\r\n\r\n'));
+
+  // And reading...
+  socket.on('data', (data) => {
+    console.log(`received ${data}`);
+    socket.close();
+  });
+
+  await socket.done();
+}
+```
+
+- UDP:
+```javascript
+import { UDP } from 'k6/x/net';
+
+export default async function () {
+  const socket = new UDP.open('192.168.1.1:9090');
+
+  await socket.write(new TextEncoder().encode('GET / HTTP/1.1\r\n\r\n'));
+}
+```
+
+- IPC:
+```javascript
+import { IPC } from 'k6/x/net';
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const socket = await IPC.open('/tmp/unix.sock');
+
+  console.log(socket.file.path); // /tmp/unix.sock
+
+  // The HTTP client supports communicating over a Unix socket.
+  const client = new Client({
+    socket: socket,
+  });
+  await client.get('http://unix/get');
+}
+```
+
+#### Client
+
+An HTTP Client is used to communicate with an HTTP server.
+
+##### Examples
+
+- Using a client with default transport settings, and making a GET request:
+```javascript
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const client = new Client();
+  const response = await client.get('https://httpbin.test.k6.io/get');
+  const jsonData = await response.json();
+  console.log(jsonData);
+}
+```
+
+- Passing a socket with custom transport settings, some HTTP options, and making a POST request:
+```javascript
+import { TCP } from 'k6/x/net';
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const socket = await TCP.open('10.0.0.10:80', { keepAlive: true });
+  const client = new Client({
+    socket: socket,
+    proxy: 'https://myproxy',
+    version: 1.1,                     // force a specific HTTP version
+    headers: { 'User-Agent': 'k6' },  // set some global headers
+  });
```

Comment:
This and the tentative HTTP/3 example below don't look very realistic, or even very user-friendly :thinking:  There are a couple of missing links here, we generally don't go from an TCP/UDP connection directly to an HTTP client. 

We want _something_ (i.e. another component) to make network connections, not to have to make them raw. And we want _something_ (also another component or a configurable part of one) to be able to negotiate which HTTP version to be used, or to be able to force that a specific HTTP version is used. Controlling these things by requiring users to make raw TCP/UDP connections doesn't seem like the way to go, especially given how complex HTTP protocol version negotiation and how it's intermingled with DNS and TLS...

Reply:
True, conceptually there's a missing "transport" abstraction between raw sockets and the high-level HTTP client. But this is an example of how both sides can be integrated when the user needs to configure socket/transport-level options. The friendlier example is the one above this, where you simply instantiate a `Client`, and make requests with it, using the default transport settings.

> we generally don't go from an TCP/UDP connection directly to an HTTP client

You can certainly talk HTTP/1 over a raw TCP socket. HTTP/2+ and TLS complicate this substantially, but ideally we would like to keep a friendly API for all versions, without introducing too many concepts.

The reason I'm hesitant to expose yet another abstraction is because the API would be even more confusing. A "transport" is uniquely a Go concept; no other JS runtime that I could find exposes it in the way Go does. For example, Deno has [`Deno.connect()`](https://deno.land/api?s=Deno.connect) and [`Deno.connectTls()`](https://deno.land/api?s=Deno.connectTls). Node has entirely separate [`net`](https://nodejs.org/api/net.html) and [`tls`](https://nodejs.org/api/tls.html) modules. That's not to say that we can't do something entirely different, but it would go against our intention of building something familiar to JS users.

That said, the following makes sense to me:

- A raw Sockets API to read/write TCP/UDP/IPC directly. Since host name resolution must happen before a connection is established, or higher level protocols are negotiated, a `lookup` hook as shown in the current examples is a good way to configure the DNS logic. This is inspired by [Node](https://nodejs.org/api/net.html#socketconnectoptions-connectlistener).

- An optional Transport API that wraps the raw socket, and implements whatever protocol or security on top. These implementations must implement the same Socket interface, so that the `http.Client` can be unaware whether it's using a raw socket connection, or e.g. a TLS connection. This would mimic Node, where [`TLSSocket` extends the base `Socket`](https://nodejs.org/api/tls.html#class-tlstlssocket), so that it can be used transparently.

  So, for example, TLS options could be specified like this:

  ```javascript
  import { TLS } from 'k6/x/net';
  const tlsSocket = await TLS.connect('10.0.0.10:443',
    { version: { min: 'ssl3.0', max: 'tls1.3'} });
  const client = new Client({
    socket: tlsSocket,
  });
  ```

- The high-level Client API would do only HTTP-specific negotiation. So if HTTP version 2 is requested, but a non-TLS connection is passed to it, it would throw an error. For that to work, an `insecure: true` option must be passed, which would negotiate `h2c`.

Let me know if you disagree with anything above.

Reply:
I fundamentally disagree with a lot of it, mostly because I don't think it will work at all, and even the parts that might work won't work well. Even if they worked, raw sockets are probably going to be the wrong abstraction 99% of the time, for anything the user may want to do with custom HTTP settings. I don't dismiss we should have support for them (though I am not sure if we should start with that), I just think there should be a layer or two between them and the HTTP `Client`.

This is why I don't think they will even work for most things. Consider even the example you provided:
```js
const tlsSocket = await TLS.connect('10.0.0.10:443',
  { version: { min: 'ssl3.0', max: 'tls1.3'} });
const client = new Client({
  socket: tlsSocket,
});
```

I see at least 4 separate issues with this small piece of code:
1. Most of the time, you need DNS to establish TLS, right? Because the certificates are for the domain names. 
2. You need to negotiate the HTTP/2 protocol version during the TLS connection establishment, so we can't easily split the TLS connection and the HTTP requests... See https://en.wikipedia.org/wiki/Application-Layer_Protocol_Negotiation
3. What happens to connection reuse? What happens if we don't want that and want a new connection for every request. It's perfectly reasonable for a single HTTP client to make multiple (potentially simultaneous) network requests to the same or to different hosts.
4. What happens if we have a redirect and we want the `Client` to follow it? :sweat_smile: 

In short, an HTTP `Client` is something that needs to be able to make its own network requests through some mechanism. You can't give it a single Socket and expect it could work just on that

Reply:
Let me focus just on one thing then:

> This passing of an already established connection is done to configure the initial connection, which the client will use instead of creating a new one.

To me this seems like a fundamentally flawed approach which can't work. It also unfortunately seems to be somewhat at the core of your proposal for handling some of the lower-level connection details. 

I already explained some of the reasons why I don't think it will work. I also don't remember **ever** seeing an approach like that, in any of the languages or libraries I've used to make HTTP requests. I don't see such an approach in any of the links you provided either:
- [node's `http.request()` has a `createConnection()` callback](https://nodejs.org/api/http.html#httprequestoptions-callback)
- [deno seems to use `Fetch`, but you can create a custom HTTP client](https://deno.land/api@v1.32.3?s=Deno.CreateHttpClientOptions&unstable=)
- [axios seems to allow you to define the `transport` and `proxy` in every request](https://github.com/axios/axios#request-config)

I may be missing something, so please share at least *one* example from *any* language or HTTP library where something like this is done, to convince me that it's a viable path forward.

It's good to have low-level network connection APIs. It's fine to somehow even allow their usage from the HTTP `Client` (e.g. by specifying a callback function that can make and return custom connections). But an HTTP client doesn't need a single network connection, it needs a mechanism to make connections, a connection factory :sweat_smile:

Reply:
I'm not claiming that what's currently here is a viable path forward. I'm expecting it will need many changes, and this is why we're discussing it. I'd like this document to be a team collaboration, rather than a traditional PR review. If you notice, the proposal author is "The k6 core team". :slightly_smiling_face:

I'm also not disagreeing with your feedback, for the most part. Some of the points you brought up are also unclear to me. So I can try to come up with a design that addresses those issues as well, and then wait more time for other concerns to pop up, or you can just suggest concrete improvements here, or—even better—push changes directly to this branch. This is not "my proposal".

This Socket <-> HTTP integration is probably my idea, so I don't have an example from similar APIs. What I was trying to avoid is the complexity of Node (notice that they also have the concept of an Agent that manages connections), and Deno doesn't really have the flexibility we need, since [its `HttpClient` is very limited](https://deno.land/api?s=Deno.HttpClient&unstable=), and is mostly a bag of options that are passed to `fetch()`. I haven't looked at Axios yet, but from a glance, that `transport` option is very unclear. :confused:

I think we can either handle this via some kind of factory hook, similar to Node's `createConnection()`, or via an event that can be emitted before creating the connection. I'm leaning towards events, as those can be applied for any advanced scenario, and are idiomatic to JS.

What I wouldn't want is to have a bunch of new abstractions, where we have to explain concepts like transport, dialer, agent, or whatever else.

Reply:
I added a `dial` hook to the `Client` constructor in 0fde82468, along with TLS options to `TCP.open()`.

I think this addresses what we discussed above, otherwise please suggest alternatives.